### ISSUE TITLE
document median(v[, region])

### ISSUE THREAD

**Issue by bjarthur** [JuliaLang/julia#12134]
i know you are busy with 0.4, but...

i find it odd that the option to specify the dimension along which to take the median is not available, as it is with the mean.

moreover, what is the rationale for not moving mode() from Distributions into Base?  this seems like a fundamental operation to me.

i stumbled across this as follows:

```
julia> data=Array{Float64,1}[]
0-element Array{Array{Float64,1},1}

julia> push!(data,rand(10))
1-element Array{Array{Float64,1},1}:
 [0.714602,0.907668,0.0968272,0.263507,0.0490918,0.0837784,0.341215,0.458432,0.0774939,0.55216]

julia> push!(data,rand(10))
2-element Array{Array{Float64,1},1}:
 [0.714602,0.907668,0.0968272,0.263507,0.0490918,0.0837784,0.341215,0.458432,0.0774939,0.55216]
 [0.784947,0.85973,0.744117,0.129139,0.572841,0.284473,0.37473,0.666274,0.384344,0.420087]     

julia> push!(data,rand(10))
3-element Array{Array{Float64,1},1}:
 [0.714602,0.907668,0.0968272,0.263507,0.0490918,0.0837784,0.341215,0.458432,0.0774939,0.55216]
 [0.784947,0.85973,0.744117,0.129139,0.572841,0.284473,0.37473,0.666274,0.384344,0.420087]     
 [0.527175,0.500384,0.500156,0.0340004,0.757666,0.850672,0.793238,0.725675,0.342043,0.275791]  

julia> mean(data)
10-element Array{Float64,1}:
 0.675575
 0.755927
 0.447034
 0.142216
 0.459866
 0.406308
 0.503061
 0.616794
 0.267961
 0.416013

julia> median(data)
ERROR: `isless` has no method matching isless(::Array{Float64,1}, ::Array{Float64,1})

julia> mode(data_traces_norm)
ERROR: mode not defined
```

**Comment by andreasnoack** [comment]
I don't see the connection between the title and the example. In the example, you are trying to find the median of a `Vector{Vector{Float64}}` and in the title you are asking about `median(v, [region])`. How would you define the median for `Vector{Vector{Float64}}`?

I don't think that mode is nearly as common as the mean variance and median. The probability theory and statistics books within my reach confirm this.

**Comment by simonster** [comment]
As @andreasnoack states, `median` doesn't work for a `Vector{Vector{Float64}}`, but neither will many operations in Base. `mean` works because the sum of vectors is well-defined, but since `median` relies on sorting elements it does not. If you want the median of a `Vector{Vector{Float64}}`, you can use `map` or a comprehension.

There is actually a `median(v, region)` function, although it's evidently missing from the docs.

**Comment by pao** [comment]
Updated title + tags to reflect current actions.

**Comment by bjarthur** [comment]
sorry for not being clear.  what i was trying to suggest, is that the API (i.e. method signatures) for mean, median, std, cov, etc. should all be identical.  it's unexpected to have ```mean(::Vector{Vector{Float64}})``` work, but not ```median(::Vector{Vector{Float64}})```.  they are all well defined if the dimensions match, which is the case in my example.

**Comment by simonster** [comment]
@bjarthur The definition of the mean (as computed by `mean`) is the sum divided by the number of elements, which is what `mean(x)` does, whether `x` is a `Vector{Float64}` or a `Vector{Vector{Float64}}`. The definition of the median is the element separating the higher half of the elements from the lower half of the elements, which is not well-defined if `x` is a `Vector{Vector{Float64}}`, because there is no natural ordering of elements of type `Vector{Float64}`.

**Comment by StefanKarpinski** [comment]
What should the median of a set of vectors be? Is it the vector that's in the middle when they are sorted lexicographically? What if there are an even number of vectors? Should it take the mean of the middle two?

**Comment by bjarthur** [comment]
oi, my bad.  i see your point now.

i will submit a PR for the median(v [,range]) documentation to make up for wasting your time.

**Comment by StefanKarpinski** [comment]
No worries. A documentation PR would be most appreciated!

**Comment by bjarthur** [comment]
the current docs say that ```median!``` is "like median", but for this to be completely true, i think the following two methods are needed, so that arrays are vectorized, as they are for ```median``` (and ```mean```).  should i include these with the PR?

```
julia> median(rand(3,5))
0.508270038530044

julia> median!(rand(3,5))
ERROR: MethodError: `median!` has no method matching median!(::Array{Float64,2})

julia> median(rand(3,5),1)
1x5 Array{Float64,2}:
 0.402373  0.52698  0.141512  0.4314  0.838239

julia> median!(rand(3,5),1)
ERROR: MethodError: `median!` has no method matching median!(::Array{Float64,2}, ::Int64)

julia> Base.median!{T}(v::AbstractArray{T}) = median!(vec(v))
median! (generic function with 3 methods)

julia> Base.median!{T}(v::AbstractArray{T}, region) = mapslices(median!, v, region)
median! (generic function with 4 methods)

julia> median!(rand(3,5))
0.6225979021603067

julia> median!(rand(3,5),1)
1x5 Array{Float64,2}:
 0.365659  0.641878  0.104584  0.148663  0.418226
```

**Comment by simonster** [comment]
`Base.median!{T}(v::AbstractArray{T}) = median!(vec(v))` is fine, but `mapslices(median!, v, region)` doesn't actually do any in-place modification. This raises a good point: I think we could just make `median{T}(v::AbstractArray{T}, region) = mapslices(median!, v, region)` and save a copy there (although it could be made more efficient still by reusing the same array for each call to `median!`).

But in general, if things were implemented optimally, I think taking the median in place would actually be slower except along the first dimension. Most of the in-place methods that operate across dimensions take the output array as the first argument and the input as the second argument, which might still be a bit faster here in some cases depending on the shape of the array. For now, I think we should just not implement or document `median!` with a region argument.

**Comment by simonster** [comment]
Fixed by #12141
